### PR TITLE
raftstorev2: avoid unnecessary snapshot after split

### LINKED ISSUE #14787: [Dynamic Regions] A slow (>1s) split in follower may cause leader to generate a snapshot
## Bug Report

As title, it's quite expensive to generate a snapshot.

### Steps to reproduce

Run `test_new_split_learner_can_not_find_leader`.

### What did you expect?

Do not generate snapshot for slow split.  

### What did happened?

A snapshot is generated.

### PR DESCRIPTION
**Pull request by overvenus**
### What is changed and how it works?

Issue Number: Close #14787

What's Changed:

```commit-message
raftstorev2: avoid unnecessary snapshot after split
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Integration test

### Release note 

```release-note
None
```

### CODE REVIEW

**Comment by SpadeA-Tang** [comment]
What about follower is pretty slow? Without the PR, the splitted regions can begin to work after received snapshot from the leader even the parent region on its store has not applied the snapshot.
Maybe we can set a deadline?

**Comment by tonyxuqqi** [comment]
So the idea is to block the first prevote or vote message if it's uninitialized. But it essentially would defeat the purpose of the v2's improvement that the un-initialized peer can still do the election so that the new region can serve traffic quickly (reduce the missing region)
Is it possible to delay the snapshot sending in leader side. The key improvement of this PR is to reduce the cross AZ traffic. If we can somehow avoid the snapshot sent in leader, then we can keep the current design's improvement.

File: components/raftstore-v2/src/operation/ready/mod.rs [hunk 1]
```diff
@@ -370,6 +370,17 @@ impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
             // It prevents cache peer(0,0) which is sent by region split.
             self.insert_peer_cache(from_peer);
         }
+
+        // Delay first message and wait for split snapshot, so that slow split
+        // does not trigger leader to send a snapshot.
+        if !self.storage().is_initialized() && is_first_message(msg.get_message()) {
+            if self.split_pending_msg_mut().replace(msg).is_some() {
```
Comment (tonyxuqqi) [inline components/raftstore-v2/src/operation/ready/mod.rs:377]: Is the new msg same as the existing one in split_pending_msg_mut()?
If it's same, then we can add an assert. If it's different, then the new message may be stepped before the old message, which could lead to correctness problem.

Comment (overvenus) [reply components/raftstore-v2/src/operation/ready/mod.rs:377]: New msgs can be different.

It's okay to step messages out of order.

File: components/raftstore-v2/src/raft/peer.rs [hunk 2]
```diff
@@ -100,6 +100,11 @@ pub struct Peer<EK: KvEngine, ER: RaftEngine> {
     // Trace which peers have not finished split.
     split_trace: Vec<(u64, HashSet<u64>)>,
     split_flow_control: SplitFlowControl,
+    /// `MsgRequestPreVote`, `MsgRequestVote` or `MsgAppend` messages from newly
```
Comment (tonyxuqqi) [inline components/raftstore-v2/src/raft/peer.rs:103]: minor: comments should not include prevote and vote. 
And split_pending_msg can be renamed as split_pending_append_msg.

**Review by tonyxuqqi — APPROVED** [review]

File: components/raftstore-v2/src/operation/ready/mod.rs [hunk 3]
```diff
@@ -370,6 +370,30 @@ impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
...
+                let dur = pending_msg.as_ref().unwrap().1.elapsed();
+                if dur < ctx.cfg.snap_wait_split_duration.0 {
+                    pending_msg.as_mut().unwrap().0 = msg;
+                    // We consider a message is too early if a message is replaced.
+                    ctx.raft_metrics.message_dropped.region_nonexistent.inc();
+                    return;
+                }
+                pending_msg.take();
+                warn!(logger, "handle first message now, split may be slow";
+                        "duration" => ?dur);
```
Comment (SpadeA-Tang) [inline components/raftstore-v2/src/operation/ready/mod.rs:393]:
```suggestion
                warn!(
                    logger, 
                    "handle first message now, split may be slow";
                    "duration" => ?dur
                );
```

**Review by SpadeA-Tang — APPROVED** [review]
rest LGTM

**Pull request merged**